tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx

```diff
@@ -66,9 +71,51 @@ private:
       }
    };
 
-   /// The index itself. Maps field values (or combinations thereof in case the index is defined for multiple fields) to
-   /// their respsective entry numbers.
-   std::unordered_map<RIndexValue, std::vector<NTupleSize_t>, RIndexValueHash> fIndex;
+   /////////////////////////////////////////////////////////////////////////////
+   /// Index for a specific entry range of the full indexed RNTuple
+   class RNTupleIndexPartition {
+      friend RNTupleIndex;
+
+   private:
+      NTupleSize_t fFirstEntry, fLastEntry;
+
+      std::unique_ptr<RPageSource> fPageSource;
+      std::vector<std::unique_ptr<RFieldBase>> fIndexFields;
+
+      /// The index itself. Maps field values (or combinations thereof in case the index is defined for multiple fields)
+      /// to their respsective entry numbers.
+      std::unordered_map<RIndexValue, std::vector<NTupleSize_t>, RIndexValueHash> fIndex;
+
+   public:
+      RNTupleIndexPartition(const RClusterDescriptor &descriptor,
+                            const std::vector<std::unique_ptr<RFieldBase>> &indexFields, const RPageSource &pageSource)
+         : fPageSource(pageSource.Clone())
+      {
+         fPageSource->Attach();
+
+         fFirstEntry = descriptor.GetFirstEntryIndex();
+         fLastEntry = fFirstEntry + descriptor.GetNEntries();
+
+         fIndexFields.reserve(indexFields.size());
+         for (const auto &field : indexFields) {
+            auto clonedField = field->Clone(field->GetFieldName());
+            CallConnectPageSourceOnField(*clonedField, *fPageSource);
```

Comment:
Now we need to clone and connect page sources and fields for every index partition / cluster, which means we probably cannot use cluster prefetching...